File: lib/workflow/submit-invoice.ts
import { z } from "zod";
import { supabaseAdmin } from "@/lib/supabase/server";
import type { InvoiceSubmission } from "@/lib/types";

export const invoiceSubmissionSchema = z.object({
  submissionId: z.string().min(1),
  source: z.enum(["email", "upload", "shared_folder", "demo_scenario"]),
  originalFileName: z.string().min(1),
  fileHash: z.string().min(1),
  mimeType: z.enum(["application/pdf", "image/png", "image/jpeg"]),
  receivedAt: z.string().min(1),
  senderEmail: z.string().email().optional(),
}) satisfies z.ZodType<InvoiceSubmission>;

export type SubmitInvoiceResult = {
  invoiceId: string;
  workflowId: string;
  status: string;
  isReplay: boolean;
};

const UNIQUE_VIOLATION = "23505";

/**
 * Idempotent workflow intake. `submissionId` is the idempotency key: a
 * replayed submission returns the existing workflow's result and never
 * creates a second row or a second job — the unique constraint on
 * invoices.submission_id is the actual enforcement point, not this
 * function's control flow, so this is race-safe under concurrent calls.
 */
export async function submitInvoice(
  submission: InvoiceSubmission
): Promise<SubmitInvoiceResult> {
  const db = supabaseAdmin();

  const { data: inserted, error: insertError } = await db
    .from("invoices")
    .insert({
      submission_id: submission.submissionId,
      source: submission.source,
      original_file_name: submission.originalFileName,
      file_hash: submission.fileHash,
      mime_type: submission.mimeType,
      received_at: submission.receivedAt,
      sender_email: submission.senderEmail ?? null,
      status: "pending",
    })
    .select("id, workflow_id, status")
    .single();

  if (insertError) {
    if (insertError.code !== UNIQUE_VIOLATION) {
      throw new Error(`submitInvoice: insert failed — ${insertError.message}`);
    }

    // Replay: the row already exists. Fetch it and log the attempt, but
    // never create a second job for it.
    const { data: existing, error: fetchError } = await db
      .from("invoices")
      .select("id, workflow_id, status")
      .eq("submission_id", submission.submissionId)
      .single();

    if (fetchError || !existing) {
      throw new Error(
        `submitInvoice: replay detected but could not load the existing row — ${fetchError?.message}`
      );
    }

    await db.from("audit_events").insert({
      workflow_id: existing.workflow_id,
      invoice_id: existing.id,
      event_id: `${existing.id}:replay:${Date.now()}`,
      stage: "replay_detected",
      label: "Replayed submission",
      detail: `submissionId ${submission.submissionId} was already processed — returned the existing workflow result without creating a new job.`,
      actor: "system",
    });

    return {
      invoiceId: existing.id,
      workflowId: existing.workflow_id,
      status: existing.status,
      isReplay: true,
    };
  }

  // New submission: create the durable job and the intake audit event.
  // If either write fails here, the invoice row still exists and is safe
  // to retry — a stuck job can always be recreated for a known invoiceId.
  await db.from("jobs").insert({
    invoice_id: inserted.id,
    job_type: "process_invoice",
    payload: { invoiceId: inserted.id, submissionId: submission.submissionId },
    status: "queued",
  });

  await db.from("audit_events").insert({
    workflow_id: inserted.workflow_id,
    invoice_id: inserted.id,
    event_id: `${inserted.id}:submission_received`,
    stage: "submission_received",
    label: "Submission received",
    detail: `${submission.source} intake, submissionId ${submission.submissionId}.`,
    actor: "system",
  });

  return {
    invoiceId: inserted.id,
    workflowId: inserted.workflow_id,
    status: inserted.status,
    isReplay: false,
  };
}

File: ledgerguard/workflow/submit_invoice.py
"""Idempotent workflow intake. `submission_id` is the idempotency key: a
replayed submission returns the existing workflow's result and never
creates a second row or a second job — the unique constraint on
invoices.submission_id is the actual enforcement point, not this function's
control flow, so this is race-safe under concurrent calls."""

import secrets
from dataclasses import dataclass
from datetime import datetime

from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from ..audit import write_audit_event
from ..models import Invoice, Job


@dataclass
class Submission:
    submission_id: str
    source: str  # email | upload | shared_folder | demo_scenario
    original_file_name: str
    file_hash: str
    mime_type: str
    received_at: datetime
    sender_email: str | None = None
    scenario_key: str | None = None
    session_token: str | None = None
    expires_at: datetime | None = None


@dataclass
class SubmitResult:
    invoice_id: str
    workflow_id: str
    status: str
    is_replay: bool


def _new_workflow_id() -> str:
    return f"wf_{secrets.token_hex(6)}"


def submit_invoice(db: Session, submission: Submission, storage_path: str) -> SubmitResult:
    workflow_id = _new_workflow_id()
    invoice = Invoice(
        submission_id=submission.submission_id,
        workflow_id=workflow_id,
        source=submission.source,
        original_file_name=submission.original_file_name,
        file_hash=submission.file_hash,
        mime_type=submission.mime_type,
        received_at=submission.received_at,
        sender_email=submission.sender_email,
        scenario_key=submission.scenario_key,
        session_token=submission.session_token,
        expires_at=submission.expires_at,
        storage_path=storage_path,
        status="received",
    )
    db.add(invoice)
    try:
        db.commit()
    except IntegrityError:
        db.rollback()
        # Replay: the row already exists. Fetch it and log the attempt, but
        # never create a second job for it.
        existing = db.scalar(select(Invoice).where(Invoice.submission_id == submission.submission_id))
        if not existing:
            raise RuntimeError("submit_invoice: replay detected but could not load the existing row") from None

        write_audit_event(
            db,
            existing.id,
            "replay_detected",
            "Replayed submission",
            f"submissionId {submission.submission_id} was already processed — returned the existing "
            "workflow result without creating a new job.",
            "system",
        )
        return SubmitResult(
            invoice_id=str(existing.id), workflow_id=existing.workflow_id, status=existing.status, is_replay=True
        )

    # New submission: create the durable job and the intake audit event. If
    # either write fails here, the invoice row still exists and is safe to
    # retry — a stuck job can always be recreated for a known invoice id.
    db.add(Job(invoice_id=invoice.id, job_type="process_invoice", status="queued"))
    db.commit()

    write_audit_event(
        db,
        invoice.id,
        "submission_received",
        "Submission received",
        f"{submission.source} intake, submissionId {submission.submission_id}.",
        "system",
    )

    return SubmitResult(invoice_id=str(invoice.id), workflow_id=workflow_id, status=invoice.status, is_replay=False)
